dispatcher/nm-dispatcher.c

```diff
@@ -575,41 +575,87 @@ dispatch_one_script (Request *request)
 	return FALSE;
 }
 
-static GSList *
-find_scripts (const char *str_action)
+static int
+_compare_basenames (gconstpointer a, gconstpointer b)
+{
+	const char *basename_a = strrchr (a, '/');
+	const char *basename_b = strrchr (b, '/');
+	int ret;
+
+	g_return_val_if_fail (basename_a, 0);
+	g_return_val_if_fail (basename_b, 0);
+
+	ret = strcmp (++basename_a, ++basename_b);
+	if (ret)
+		return ret;
+
+	return strcmp (a, b);
+}
+
+static void
+_find_scripts (GHashTable *scripts, const char *base, const char *subdir)
 {
-	GDir *dir;
 	const char *filename;
-	GSList *sorted = NULL;
+	gs_free char *dirname = NULL;
 	GError *error = NULL;
-	const char *dirname;
+	GDir *dir;
 
-	if (   strcmp (str_action, NMD_ACTION_PRE_UP) == 0
-	    || strcmp (str_action, NMD_ACTION_VPN_PRE_UP) == 0)
-		dirname = NMD_SCRIPT_DIR_PRE_UP;
-	else if (   strcmp (str_action, NMD_ACTION_PRE_DOWN) == 0
-	         || strcmp (str_action, NMD_ACTION_VPN_PRE_DOWN) == 0)
-		dirname = NMD_SCRIPT_DIR_PRE_DOWN;
-	else
-		dirname = NMD_SCRIPT_DIR_DEFAULT;
+	dirname = g_build_filename (base, "dispatcher.d", subdir, NULL);
 
 	if (!(dir = g_dir_open (dirname, 0, &error))) {
 		g_message ("find-scripts: Failed to open dispatcher directory '%s': %s",
 		           dirname, error->message);
 		g_error_free (error);
-		return NULL;
+		return;
 	}
 
 	while ((filename = g_dir_read_name (dir))) {
-		char *path;
+		g_hash_table_insert (scripts,
+		                     g_strdup (filename),
+		                     g_build_filename (dirname, filename, NULL));
+	}
+
+	g_dir_close (dir);
+}
+
+static GSList *
+find_scripts (const char *str_action)
+{
+	GSList *script_list = NULL;
+	GHashTable *scripts = NULL;
+	GHashTableIter iter;
+	const char *subdir = NULL;
+	char *path;
+	char *filename;
+
+	if (   strcmp (str_action, NMD_ACTION_PRE_UP) == 0
+	    || strcmp (str_action, NMD_ACTION_VPN_PRE_UP) == 0)
+		subdir = "pre-up.d";
+	else if (   strcmp (str_action, NMD_ACTION_PRE_DOWN) == 0
+	         || strcmp (str_action, NMD_ACTION_VPN_PRE_DOWN) == 0)
+		subdir = "pre-down.d";
+
+	scripts = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+
+	_find_scripts (scripts, NMLIBDIR, subdir);
+	_find_scripts (scripts, NMCONFDIR, subdir);
```

Comment:
ok, this puts all files into the scripts hash, shadowing files according to the same filename. But only below, it checks for the file permissions. I think it's correct that a non-executable file (owned by root) shadows an executable script. But if the shadowing file is a directory or owned by the wrong user or writable to non-root, I think it should not shadow a script. Or at least, it's a question that should be considered before agreeing that shadowing only happens based on the filename alone, without regard of owner/permissions/type.

Reply:
I can't see why anything but file name would decide whether something shadows something else. It just takes extra effort for no benefit.

Also the permission check is just an extra safety net. Nobody but root can actually install the scripts anyway.